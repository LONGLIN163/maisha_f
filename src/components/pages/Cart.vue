<template>

    <div>
        <div class="navbar-div">
            <van-nav-bar
                title="Cart"
                left-text="Back"
                left-arrow
                @click-left="onClickLeft"
            ></van-nav-bar>
            <!--***van-nav-bar has to be a close tag, otherwise will have error occurs in testing***-->
        </div>

        <div class="card-title">
            <van-button size="small" type="danger" @click="clearCart" plain>Empty Cart</van-button>
        </div>
        <div class="cart-list">
            <div class="cart-row" v-for="(item,index) in cartInfo" :key="index">
                <div class="cart-img"><img :src="item.image" width="100%" /></div>
                <div class="cart-text">
                    <div class="cart-goods-name">{{item.Name}}</div>

                        <div class="cart-control">
                        <van-stepper v-model="item.count" />
                    </div>
                </div>
                <div class="cart-goods-price">
                    <div>
                            {{item.price |moneyFilter}} €
                    </div>
                    <div>
                        x{{item.count}}
                    </div>
                    <div class="allPrice">
                        {{item.price*item.count | moneyFilter}} €
                    </div>

                </div>
         
            </div>
        </div>
        <div class="totalMoney">

            Total: {{totalMoney | moneyFilter}} €

        </div>
    </div>
    
</template>

<script>
import {toMoney} from '@/filters/moneyFilter.js'
export default {
    data() {
        return {
            cartInfo: [],     //product item in the shopping cart
            isEmpty : false , 
        }
    },
    created(){
        this.getCartInfo()
    },
    computed:{
        totalMoney(){
            let allMoney = 0
            this.cartInfo.forEach((item) => {
               allMoney += item.price*item.count
            });

            localStorage.cartInfo=JSON.stringify(this.cartInfo)

            return allMoney
        }
    },
    methods: {
        getCartInfo() {
            if(localStorage.cartInfo){
                this.cartInfo = JSON.parse(localStorage.cartInfo) 
            }
            console.log(' this.cartInfo:'+JSON.stringify( this.cartInfo))
            this.isEmpty = this.cartInfo.length>0 ? true : false
        },
        clearCart(){
            localStorage.removeItem('cartInfo')
            this.cartInfo=[]
        },
        onClickLeft(){ // go to last page
            this.$router.go(-1)
        }
    },
    filters:{
        moneyFilter(money){
            return toMoney(money)
        }  
    }
}
</script>

<style lang="scss" scoped>
.card-title{
    height: 2rem;
    line-height:2rem;
    background-color: $color-background-sub;
    border-bottom:$color-divider;
    padding:5px;
    text-align: right;
}
.cart-list{
    background-color: $color-background-sub;
    .cart-row{
        display: flex;
        flex-direction: row;
        flex-wrap:nowrap;
        padding:0.5rem;
        font-size:0.85rem;
        border-bottom:$color-divider;
        .cart-img{
            flex:6;
        }
        .cart-text{
            flex:14;
            padding-left:10px;
            .cart-control{
                padding-top: 10px;
            }
        }
        .cart-goods-price{
            flex:4;
            text-align: right;
        }
    }
}
.totalMoney{
    text-align: right;
    background-color: $color-secondary;
    border-bottom:$color-divider;
    padding: 5px;
}
</style>