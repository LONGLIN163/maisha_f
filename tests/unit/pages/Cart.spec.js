import { mount, createLocalVue} from '@vue/test-utils'
import Cart from '@/components/pages/Cart.vue'
import {cartInfo} from '../../mockData'

const Vant=require("vant")
const localVue = createLocalVue();
localVue.use(Vant); 

describe('Cart', () => {
  
  let wrapper;

   // mock getCartInfo()
   const getCartInfo = jest.spyOn(Cart.methods, 'getCartInfo')
   .mockImplementation(() => {
     console.log("mock getCartInfo implementation")
   })
  
   // mock totalMoney()
   const totalMoney = jest.spyOn(Cart.computed, 'totalMoney')
   .mockImplementation(() => {
     console.log("mock totalMoney implementation")
   })
  
   // mock clearCart()
   const clearCart = jest.spyOn(Cart.methods, 'clearCart')
   .mockImplementation(() => {
     console.log("mock clearCart implementation")
   })
  
  beforeEach(() => {
    wrapper = mount(Cart,{
      localVue
    })
    wrapper.setData({ cartInfo })
  })
  
  
  it('renders Cart', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('getCartInfo() and totalMoney() should be called', () => {
    expect(getCartInfo).toHaveBeenCalled()
    expect(totalMoney).toHaveBeenCalled()
  })

  it('clearCart should be called', async () => {
    await wrapper.find('[data-test="clearCart"]').trigger('click');
    expect(clearCart).toHaveBeenCalled()
  })

   // ******van-stepper can be test in e2e******
  
})