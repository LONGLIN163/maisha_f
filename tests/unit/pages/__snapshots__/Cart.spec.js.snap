// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart renders Cart 1`] = `
<div>
  <div class="navbar-div">
    <div class="van-nav-bar van-hairline--bottom">
      <div class="van-nav-bar__content">
        <div class="van-nav-bar__left"><i class="van-icon van-icon-arrow-left van-nav-bar__arrow">
            <!---->
          </i><span class="van-nav-bar__text">Back</span></div>
        <div class="van-nav-bar__title van-ellipsis">Cart</div>
      </div>
    </div>
  </div>
  <div class="card-title"><button data-test="clearCart" class="van-button van-button--danger van-button--small van-button--plain">
      <div class="van-button__content"><span class="van-button__text">Empty Cart</span></div>
    </button></div>
  <div class="cart-list">
    <div class="cart-row">
      <div class="cart-img"><img src="http://images.baixingliangfan.cn/shopGoodsImg/20171225/20171225111353_2166.jpg" width="100%"></div>
      <div class="cart-text">
        <div class="cart-goods-name">product name</div>
        <div class="cart-control">
          <div class="van-stepper"><button type="button" class="van-stepper__minus van-stepper__minus--disabled"></button><input type="text" role="spinbutton" inputmode="decimal" aria-valuemax="Infinity" aria-valuemin="1" aria-valuenow="1" class="van-stepper__input"><button type="button" class="van-stepper__plus"></button></div>
        </div>
      </div>
      <div class="cart-goods-price">
        <div>
          6.00 €
        </div>
        <div>
          x1
        </div>
        <div class="allPrice">
          6.00 €
        </div>
      </div>
    </div>
    <div class="cart-row">
      <div class="cart-img"><img src="http://images.baixingliangfan.cn/shopGoodsImg/20171226/20171226162234_480.jpg" width="100%"></div>
      <div class="cart-text">
        <div class="cart-goods-name">product name</div>
        <div class="cart-control">
          <div class="van-stepper"><button type="button" class="van-stepper__minus van-stepper__minus--disabled"></button><input type="text" role="spinbutton" inputmode="decimal" aria-valuemax="Infinity" aria-valuemin="1" aria-valuenow="1" class="van-stepper__input"><button type="button" class="van-stepper__plus"></button></div>
        </div>
      </div>
      <div class="cart-goods-price">
        <div>
          12.00 €
        </div>
        <div>
          x1
        </div>
        <div class="allPrice">
          12.00 €
        </div>
      </div>
    </div>
  </div>
  <div class="totalMoney">

    Total: 0.00 €

  </div>
</div>
`;
